src/dlsproto/client/request/internal/GetRange.d

```diff
@@ -744,41 +789,60 @@ private scope class GetRangeHandler
-                    Message(MessageType_v2.Finished));
+                    Message(MessageType_v2.Finished),
+                    Signal(this.outer.TimeoutSignal));
 
-                switch (msg.type)
+                with (event.Active) switch (event.active)
                 {
-                    case MessageType_v2.Records:
-                        Const!(void)[] received_record_batch;
-                        size_t uncompressed_batch_size;
+                    case message:
+                        auto msg = event.message;
+                        switch (msg.type)
+                        {
+                            case MessageType_v2.Records:
+                                Const!(void)[] received_record_batch;
+                                size_t uncompressed_batch_size;
 
-                        this.outer.conn.message_parser.parseBody(
-                            msg.payload, uncompressed_batch_size,
-                            received_record_batch
-                        );
+                                this.outer.conn.message_parser.parseBody(
+                                    msg.payload, uncompressed_batch_size,
+                                    received_record_batch
+                                );
 
-                        (*uncompressed_batch).length = uncompressed_batch_size;
-                        enableStomping(*uncompressed_batch);
+                                (*uncompressed_batch).length = uncompressed_batch_size;
+                                enableStomping(*uncompressed_batch);
 
-                        this.lzo.uncompress(received_record_batch,
-                               *uncompressed_batch);
-                        this.record_stream.addRecords(*uncompressed_batch);
-                        break;
+                                this.lzo.uncompress(received_record_batch,
+                                       *uncompressed_batch);
+                                this.record_stream.addRecords(*uncompressed_batch);
+                                break;
 
-                    case MessageType_v2.Stopped:
-                        this.record_stream.stop();
-                        return;
+                            case MessageType_v2.Stopped:
+                                this.record_stream.stop();
+                                return;
 
-                    case MessageType_v2.Finished:
+                            case MessageType_v2.Finished:
+                                finished = true;
+                                break;
+
+                            default:
+                                assert(false);
+                        }
+                        break;
+                    case signal:
+                        verify(event.signal.code == this.outer.TimeoutSignal);
+                        // Got a message from timer, give it chance to exit
+                        // before dismantling this RoC
```

Comment:
What exactly was the problem here? I wonder if there's a way we can make the API simpler so that this isn't required. This caveat should at least be mentioned in the timer classes.

Reply:
It's related to the behaviour of this method, right? https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/request/RequestEventDispatcher.d#L465-L490

>Any fiber which is awaiting this signal will be resumed in the next eventLoop cycle.

It looks like the documentation there is misleading, by the way. It attempts to resume the waiting fibers right away, and only if they can't be does it queue the signal. (As far as I recall.)

Reply:
Yes, the problem is that timer immediately can resume the fiber (and not in the next eventLoop cycle - that's misleading). 

So, what happens here is:

1. epoll's cycle starts the timer's handler
2. We dispatch the signal and resume the fiber directly - the timer's routine gets blocked
3. We then react to this timeout signal and dismantle the entire request
4. The timer's callback (from the main fiber) now continues, goes back, _but_ it runs invariants on the objects that no longer exist!
5. 💥

Reply:
> What exactly was the problem here? I wonder if there's a way we can make the API simpler so that this isn't required. This caveat should at least be mentioned in the timer classes.

I'm not sure timer's API is in fault here - I was trying to add `delayedSignal` to `RequestEventDispatcher`, but it didn't work for the reasons I didn't look much into, because I can't afford swarm update in the minor release anyway :-/. However, I think that's what we need - some way of delayed sending (if possible, the fiber just didn't want to wake up, I think it will need a bit more of refactoring of request event dispatcher then simply to queue the signal).

We have the same issue in ocean's Task system - where it is not the caller that needs to run `processEvents` but the running task.

Reply:
I see. I'd expand the comment to:
"Got a message from timer. Yield in order to allow it to unregister itself from epoll, before this RoC exits (the timer will be destroyed when this happens)."